core/src/main/java/org/elasticsearch/tasks/TaskManager.java

```diff
@@ -39,6 +41,8 @@
 
     private final AtomicLong taskIdGenerator = new AtomicLong();
 
+    private final Collection<TaskManagerListener> listeners = new CopyOnWriteArrayList<>();
```

Comment:
Because TaskManager is created by TransportService, in order to pass an unmodifiable map, we will have to essentially create and pass any listeners to the TransportService during its creation. This will limit an ability of a service that wants to use TransportService or ClusterService to register itself as a listener because of the circular dependency:

Service <-- TransportService <-- ClusterService <-- Service